db/migrate/20180410092453_policy_log.rb

```diff
@@ -0,0 +1,87 @@
+Sequel.migration do
+  tables = %i(roles role_memberships resources permissions annotations)
+
+  up do
+    execute """
+      CREATE TYPE policy_log_op AS ENUM ('INSERT', 'DELETE', 'UPDATE');
+      CREATE TYPE policy_log_kind AS ENUM #{literal tables.map(&:to_s)};
+      CREATE EXTENSION IF NOT EXISTS hstore;
+    """
+
+    key = %i(policy_id version)
+    create_table :policy_log do
+      String :policy_id, null: false
+      Integer :version, null: false
+      foreign_key key, :policy_versions, on_delete: :cascade
+      index key
+
+      column :operation, :policy_log_op, null: false
+      column :kind, :policy_log_kind, null: false
+      column :subject, :hstore, null: false
+      column :at, :timestamptz, null: false, default: Sequel.function(:clock_timestamp)
+    end
+
+    tables.each do |table|
+      # find the primary key of the table
+      primary_key = schema(table).select{|x,s|s[:primary_key]}.map(&:first).map(&:to_s).pg_array
+      execute """
+        CREATE OR REPLACE FUNCTION policy_log_#{table}() RETURNS TRIGGER AS $$
+          DECLARE
+            subject #{table};
+            current policy_versions;
+          BEGIN
+            IF (TG_OP = 'DELETE') THEN
+              subject := OLD;
+            ELSE
+              subject := NEW;
+            END IF;
+            current = current_policy_version();
+            IF current.resource_id = subject.policy_id THEN
+              INSERT INTO policy_log(
+                policy_id, version,
+                operation, kind,
+                subject)
+              SELECT
+                current.resource_id, current.version,
+                TG_OP::policy_log_op, '#{table}'::policy_log_kind,
+                slice(hstore(subject), #{literal primary_key})
+              ;
+            ELSE
+              RAISE WARNING 'modifying data outside of policy load: %', subject.policy_id;
```

Comment:
Is this a warning or an error? Is it here so that mucking with the tables isn't blocked by auditing? Do we want to enforce that table modification can only be performed if there is an open policy_version?

If we were sure that every change was audited, then the current state of the audited tables could be built from scratch by replaying the audit. If we allow unaudited access, then we can't make that guarantee. 

I could imagine a switch that allows unaudited writes in dev and test; without that switch in place, this line would be an error and abort the transaction.

Reply:
I left it as a warning for now since it happens in tests a bit and I wasn't sure if it isn't happening elsewhere.